historical_surveillance/views.py

```diff
@@ -0,0 +1,328 @@
+    districts_names = district_df['District_Name']
+    return render(request, 'enrollment.html', {'form': form, 'district_names': districts_names})
+
+
+# This function controls the total enrollment and school capacity form, processes the form data and submits
+# it to the database
+def enroll_class(request):
+    form = AggregateEnrollmentForms()
+    if request.method == 'POST':
+        form = AggregateEnrollmentForms(request.POST)
+        if form.is_valid():
+            school_form = form.cleaned_data['name_of_school']
+            academic_year_form = form.cleaned_data['academic_year']
+            district_form = form.cleaned_data['district_of_school']
+            if not AggregateEnrollment.objects.filter(name_of_school=school_form, academic_year=academic_year_form,
+                                                      district_of_school=district_form).exists():
+                form.save()
+                return redirect("/")
+        else:
+            messages.add_message(request, messages.INFO, 'Data has been entered previously')
+
+    return render(request, 'enroll_class.html', {'form': form})
+
+
+# This view controls the enrollment / Capacity views table
+def enrolled(request):
+    data = AggregateEnrollment.objects.all()
+
+    stu = {
+        "name_of_school": data
+    }
+    return render(request, 'enrolled.html', stu)
+
+#change this to get a form to select the district and pass it as a parameter to present filter the table and present
+# the data for each district
+def enrolled_trend_district_1(request):
+    error_message = None
+    district_df = pd.DataFrame(District.objects.all().values())
+    districts_names = district_df['District_Name']
+    if district_df.shape[0] > 0:
+        if request.method == 'POST':
+            district_selected = request.POST.get('district_name', False)
+            if not district_selected:
+                error_message = "Please select a chart type and school"
+            else:
+                if district_selected is not False:
+                    data = AggregateEnrollment.objects.all().filter(district_of_school=district_selected)
+                stu = {
+                    "error_message": error_message,
+                    "districts_names": districts_names,
+                    "district_selected" : district_selected,
+                    "data" : data,
+                }
+                return render(request, 'district_1.html', stu)
+        else:
+            return render(request, 'district_1.html', {"districts_names":districts_names})
+
+
+
+    else:
+        error_message = "No records found for the selected district"
+
+
+
+def district_1_trend(request):
+    error_message = None
+    graph = None
+
+    district_df = pd.DataFrame(District.objects.all().values())
+    districts_lists = district_df['District_Name']
+
+    schools_df = pd.DataFrame(AggregateEnrollment.objects.all().values())
+    schools = schools_df['name_of_school'].unique()
```

Comment:
this can be achieved doing:

`AggregateEnrollment.objects.values('name_of_school').distinct()`

no data frame needed

Reply:
The data frame is used beyond just checking if there is data. It is from that date frame that the unique names of schools are filtered.

Reply:
Querying all data in a table and putting it into memory is expensive. I see 2 uses for the data frame here:

1. Check if the table has data: `AggregateEnrollment.objects.count() > 0`
2. Get all unqiue school names: `AggregateEnrollment.objects.values('name_of_school').distinct()` or if you need a. list
`AggregateEnrollment.objects.values_list('name_of_school', flat=True).distinct()`

Both of these queries do not load all the data from the table in memory they only fetch what is needed and therefore it is more efficient this way.